### PR TITLE
Only delete storage pools that are not in use.

### PR DESCRIPTION
**Pull request by Veebers**
## Description of change

Add a check when deleting a storage pool that it isn't in use. If it is in use do not delete that pool.
Note this PR builds off of: https://github.com/juju/juju/pull/9682

## QA steps

Unit tests included.
Manual test: 
  - Follow the steps found here: https://discourse.jujucharms.com/t/juju-kubernetes-and-microk8s/226
    - At the point of creating the ```operator-storage```, run: ```juju delete-storage-pool operator-storage```
    - Run: ```juju list-storage-pools``` to show its gone
    - Create the pool again and continue with the instructions
    - Once the application has been deployed attempt to delete the storage pool
      - ```juju delete-storage-pool mariadb-pv``` must result in an error 'ERROR pool "mariadb-pv" in use'
      - ```juju delete-storage-pool operator-storage``` must result in an error 'ERROR pool "operator-storage" in use'

## Documentation changes
n/a

## Bug reference
n/a

### CODE REVIEW

File: apiserver/facades/client/storage/storage.go [hunk 1]
```diff
@@ -26,6 +26,10 @@ import (
 	"github.com/juju/juju/storage/poolmanager"
 )
 
+const (
+	operatorStorage = "operator-storage"
```
Comment (wallyworld) [inline apiserver/facades/client/storage/storage.go]: There's already a const OperatorStoragePoolName in caas package

File: apiserver/facades/client/storage/storage.go [hunk 2]
```diff
@@ -1155,6 +1168,31 @@ func (a *StorageAPI) DeletePool(p params.StoragePoolDeleteArgs) (params.ErrorRes
 	return results, nil
 }
 
+func (a *StorageAPI) storagePoolsInUse(pools []params.StoragePoolDeleteArg) (map[string]bool, error) {
```
Comment (wallyworld) [inline apiserver/facades/client/storage/storage.go]: Use a Set not a map[string]bool.

File: apiserver/facades/client/storage/storage.go [hunk 3]
```diff
@@ -1145,7 +1149,16 @@ func (a *StorageAPI) DeletePool(p params.StoragePoolDeleteArgs) (params.ErrorRes
 	if err := a.checkCanWrite(); err != nil {
 		return results, errors.Trace(err)
 	}
+	poolsInUse, err := a.storagePoolsInUse(p.Pools)
+	if err != nil {
+		return results, errors.Trace(err)
+	}
+
 	for i, pool := range p.Pools {
+		if poolsInUse[pool.Name] {
```
Comment (wallyworld) [inline apiserver/facades/client/storage/storage.go]: This is not really what we want - we want the Delete() call below to return an error if the specific pool being deleted is in use

**Comment by mitechie** [comment]
We shouldn't have any CLI with delete. This would be remove or destroy from a UX perspective.

File: state/storage.go [hunk 4]
```diff
@@ -295,6 +298,34 @@ func (sb *storageBackend) AllStorageInstances() ([]StorageInstance, error) {
 	return out, nil
 }
 
+// DeleteStoragePool removes a pool only if its not currently in use
+func (sb *storageBackend) DeleteStoragePool(poolName string) error {
+	storageCollection, closer := sb.mb.db().GetCollection(storageInstancesC)
+	defer closer()
+
+	var inUse bool
+	if sb.modelType == ModelTypeCAAS && poolName == caas.OperatorStoragePoolName {
```
Comment (wallyworld) [inline state/storage.go]: Add TODO to say that we need to change the data model to support doing this as an assert

File: state/storage_test.go [hunk 5]
```diff
@@ -1346,6 +1383,54 @@ func (s *StorageStateSuite) TestNewModelDefaultPools(c *gc.C) {
...
+
+	err = s.storageBackend.DeleteStoragePool(poolName)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Assert(err, jc.ErrorIsNil)
+	_, err = s.pm.Get(poolName)
+	c.Assert(err, gc.ErrorMatches, fmt.Sprintf("pool %q not found", poolName))
+}
+
+func (s *StorageStateSuiteCaas) TestDeleteStoragePoolErrors(c *gc.C) {
```
Comment (wallyworld) [inline state/storage_test.go]: TestDeleteStoragePoolInUse

File: state/storage_test.go [hunk 6]
```diff
@@ -658,6 +664,37 @@ func (s *StorageStateSuite) TestAllStorageInstances(c *gc.C) {
...
+	_, err := s.pm.Get(poolName)
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.storageBackend.DeleteStoragePool(poolName)
+	c.Assert(err, jc.ErrorIsNil)
+	_, err = s.pm.Get(poolName)
+	c.Assert(err, gc.ErrorMatches, fmt.Sprintf("pool %q not found", poolName))
+}
+
+func (s *StorageStateSuite) TestDeleteStoragePoolErrors(c *gc.C) {
```
Comment (wallyworld) [inline state/storage_test.go]: TestDeleteStoragePoolInUse

**Comment by Veebers** [comment]
$$merge$$

**Pull request merged**